File: idt_front/src/types/blueprint.ts
// golden-sample-blueprint Design §4.2 — 백엔드 src/interfaces/schemas/blueprint.py 와 동기.
// PatternKind/SlotKind 값은 백엔드 StrEnum value (wire 계약 — 변경 금지).

export const PATTERN_KINDS = [
  'cover',
  'toc',
  'section_lead',
  'text',
  'chart_with_notes',
  'table',
  'two_column',
  'image_with_notes',
  'closing',
  'unknown',
] as const;
export type PatternKind = (typeof PATTERN_KINDS)[number];

export const SLOT_KINDS = ['title', 'text', 'bullets', 'table', 'chart', 'image', 'footer'] as const;
export type SlotKind = (typeof SLOT_KINDS)[number];

export const ASSET_KINDS = ['logo', 'decoration', 'cover'] as const;
export type AssetKind = (typeof ASSET_KINDS)[number];

export const SOURCE_KINDS = ['pdf', 'pptx'] as const;
export type SourceKind = (typeof SOURCE_KINDS)[number];

export type BlueprintStatus = 'active' | 'inactive';

export const PATTERN_KIND_LABELS: Record<PatternKind, string> = {
  cover: '표지',
  toc: '목차',
  section_lead: '섹션 리드',
  text: '본문',
  chart_with_notes: '차트+해설',
  table: '표',
  two_column: '2단',
  image_with_notes: '이미지+해설',
  closing: '결론',
  unknown: '미분류',
};

export const SLOT_KIND_LABELS: Record<SlotKind, string> = {
  title: '제목',
  text: '텍스트',
  bullets: '불릿',
  table: '표',
  chart: '차트',
  image: '이미지',
  footer: '푸터',
};

export const ASSET_KIND_LABELS: Record<AssetKind, string> = {
  logo: '로고',
  decoration: '장식',
  cover: '표지',
};

/** 클라이언트 가드 — 백엔드 라우터/UseCase 와 동일 숫자 (Design §4.2 / §7) */
export const BLUEPRINT_LIMITS = {
  upload_max_bytes: 30 * 1024 * 1024,
  name_max: 100,
  description_max: 500,
  max_pages_default: 60,
  max_assets: 20,
} as const;

export const REQUIRED_PALETTE_KEYS = ['primary', 'accent1', 'text', 'bg'] as const;
export const REQUIRED_SIZE_KEYS = ['h1', 'h2', 'body', 'caption'] as const;
/** schema v2 옵션 크기 역할 (blueprint-style-fidelity DR-4) — 없으면 서버가 폴백 */
export const OPTIONAL_SIZE_KEYS = ['h3', 'subtitle'] as const;
export const CURRENT_BLUEPRINT_SCHEMA_VERSION = 2;
export type SlotAlign = 'left' | 'center' | 'right';

// ── blueprint 구조 (blueprint_to_dict 와 동형) ─────────────────

export interface RelBox {
  x: number;
  y: number;
  w: number;
  h: number;
}

export interface BlueprintSlot {
  id: string;
  kind: SlotKind;
  box: RelBox;
  role: string;
  max_chars: number | null;
  max_rows: number | null;
  asset_id: string | null;
  /** schema v2 — v1 응답에는 없을 수 있음 (서버 기본 'left') */
  align?: SlotAlign;
}

/** schema v2 — 서버 계산 장식 도형. LLM 이 채우지 않음 (읽기 전용 라운드트립). */
export interface BlueprintDecoration {
  id: string;
  shape: 'rect';
  box: RelBox;
  fill: string;
  line: string | null;
}

export interface BlueprintPattern {
  id: string;
  kind: PatternKind;
  slots: BlueprintSlot[];
  background: string | null;
  sample_page: number;
  notes: string;
  /** schema v2 */
  decorations?: BlueprintDecoration[];
}

export interface TableStyle {
  header_bg: string;
  header_text: string;
  border: string;
  zebra: boolean;
}

export interface HeaderFooter {
  logo_asset_id: string | null;
  page_number_format: string;
  footer_text: string;
  /** schema v2 */
  footer_box?: RelBox | null;
  page_number_box?: RelBox | null;
  footer_color?: string | null;
}

export interface BlueprintStyle {
  slide_size: number[];
  fonts: Record<string, string>;
  sizes: Record<string, number>;
  palette: Record<string, string>;
  table_style: TableStyle;
  header_footer: HeaderFooter;
  /** schema v2 — 비표지 전 슬라이드 공통 장식 */
  common_decorations?: BlueprintDecoration[];
}

export interface NarrativeSection {
  role: string;
  pattern_ids: string[];
  guidance: string;
}

export interface BlueprintNarrative {
  sections: NarrativeSection[];
  tone: string;
  language: string;
}

export interface BlueprintAsset {
  id: string;
  kind: AssetKind;
  mime: string;
  width: number;
  height: number;
  sha256: string;
  box: RelBox;
  adopted: boolean;
  /** schema v2 — 표지에서의 위치 (logo 전용) */
  cover_box?: RelBox | null;
}

/** 요청용 draft (타임스탬프 없음) — POST/PUT body.draft */
export interface BlueprintDraft {
  id: string;
  name: string;
  description: string;
  source_kind: SourceKind;
  page_count: number;
  schema_version: number;
  style: BlueprintStyle;
  patterns: BlueprintPattern[];
  narrative: BlueprintNarrative;
  assets: BlueprintAsset[];
  font_mapping: Record<string, string>;
  warnings: string[];
  status: BlueprintStatus;
}

export interface BlueprintResponse extends BlueprintDraft {
  created_at: string;
  updated_at: string;
}

export interface BlueprintSummary {
  id: string;
  name: string;
  source_kind: SourceKind;
  page_count: number;
  pattern_count: number;
  status: BlueprintStatus;
  updated_at: string;
}

export interface AssetPreview extends BlueprintAsset {
  thumbnail_b64: string | null;
  /** 추출 응답에만 포함 — 저장(POST) 시 되돌려 보내는 원본 바이트 */
  data_b64: string | null;
}

export interface BlueprintExtractResponse {
  draft: BlueprintResponse;
  assets: AssetPreview[];
  page_thumbnails: (string | null)[];
  classification: { succeeded: number; failed: number };
  timings_ms: Record<string, number>;
}

export interface BlueprintCreateRequest {
  draft: BlueprintDraft;
  /** 추출 응답 assets[].data_b64 를 그대로 되돌려 보낸다 */
  assets: { id: string; data_b64: string }[];
}

export interface BlueprintUpdateRequest {
  draft: BlueprintDraft;
}

export interface FontsResponse {
  installed: string[];
  default: string;
}

export interface BlueprintOption {
  id: string;
  name: string;
}

export interface BlueprintOptionsResponse {
  items: BlueprintOption[];
}

File: idt/src/interfaces/schemas/blueprint.py
"""golden-sample-blueprint API 스키마 (Design §4.2).

요청/응답은 domain serialization(blueprint_to_dict/from_dict) 위에서 동작 —
BlueprintPayload 는 타임스탬프를 제외한 blueprint dict 를 그대로 전달하며 VO 검증이
단일 출처(ValueError → 400).
"""

from __future__ import annotations

import base64
from collections.abc import Callable
from datetime import UTC, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from src.application.blueprint.extraction_use_case import ExtractionOutcome
from src.domain.blueprint.font_normalization import normalize_blueprint_fonts
from src.domain.blueprint.serialization import blueprint_from_dict, blueprint_to_dict
from src.domain.blueprint.value_objects import (
    CURRENT_SCHEMA_VERSION,
    BlueprintAsset,
    DocumentBlueprint,
)


class _Strict(BaseModel):
    model_config = ConfigDict(extra="forbid")


class RelBoxSchema(_Strict):
    x: float
    y: float
    w: float
    h: float


class SlotSchema(_Strict):
    id: str
    kind: str
    box: RelBoxSchema
    role: str = ""
    max_chars: int | None = None
    max_rows: int | None = None
    asset_id: str | None = None
    align: Literal["left", "center", "right"] = "left"  # schema v2


class DecorationSchema(_Strict):
    """schema v2 — 서버 계산 장식 도형 (blueprint-style-fidelity §4.1)."""

    id: str
    shape: Literal["rect"] = "rect"
    box: RelBoxSchema
    fill: str
    line: str | None = None


class PatternSchema(_Strict):
    id: str
    kind: str
    slots: list[SlotSchema]
    background: str | None = None
    sample_page: int = 0
    notes: str = ""
    decorations: list[DecorationSchema] = []  # schema v2


class TableStyleSchema(_Strict):
    header_bg: str
    header_text: str
    border: str
    zebra: bool
    # blueprint-render-style-fidelity — 선택 필드 (기본값 = 현행 동작)
    zebra_bg: str = "#F3F4F6"
    border_width_pt: float = 0.0


class HeaderFooterSchema(_Strict):
    logo_asset_id: str | None = None
    page_number_format: str = ""
    footer_text: str = ""
    footer_box: RelBoxSchema | None = None  # schema v2
    page_number_box: RelBoxSchema | None = None  # schema v2
    footer_color: str | None = None  # schema v2


class StyleSchema(_Strict):
    slide_size: list[float]
    fonts: dict[str, str]
    sizes: dict[str, float]
    palette: dict[str, str]
    table_style: TableStyleSchema
    header_footer: HeaderFooterSchema
    common_decorations: list[DecorationSchema] = []  # schema v2
    # blueprint-render-style-fidelity — 선택 필드 (기본값 = 현행 동작)
    body_line_spacing: float = 1.0
    body_space_after_pt: float = 0.0
    chart_label_size_pt: float = 0.0
    chart_label_bold: bool = True


class NarrativeSectionSchema(_Strict):
    role: str
    pattern_ids: list[str]
    guidance: str = ""


class NarrativeSchema(_Strict):
    sections: list[NarrativeSectionSchema]
    tone: str = ""
    language: str = "ko"


class AssetSchema(_Strict):
    id: str
    kind: Literal["logo", "decoration", "cover"]
    mime: str
    width: int
    height: int
    sha256: str
    box: RelBoxSchema
    adopted: bool = True
    cover_box: RelBoxSchema | None = None  # schema v2


class BlueprintPayload(_Strict):
    """요청용 blueprint (타임스탬프 없음)."""

    id: str
    name: str = Field(max_length=100)
    description: str = Field(default="", max_length=500)
    source_kind: Literal["pdf", "pptx"]
    page_count: int
    schema_version: int = CURRENT_SCHEMA_VERSION
    style: StyleSchema
    patterns: list[PatternSchema]
    narrative: NarrativeSchema
    assets: list[AssetSchema] = []
    font_mapping: dict[str, str] = {}
    warnings: list[str] = []
    status: Literal["active", "inactive"] = "active"

    def to_domain(self) -> DocumentBlueprint:
        now = datetime.now(UTC).isoformat()
        data = self.model_dump()
        data.update(created_at=now, updated_at=now)
        # Design Ref: blueprint-font-mapping-migration DR-1 — 쓰기 경계 정규화(FR-03).
        return normalize_blueprint_fonts(blueprint_from_dict(data))


class AssetUpload(_Strict):
    id: str
    data_b64: str

    def decode(self) -> bytes:
        return base64.b64decode(self.data_b64)


class BlueprintCreateRequest(_Strict):
    draft: BlueprintPayload
    assets: list[AssetUpload] = []


class BlueprintUpdateRequest(_Strict):
    draft: BlueprintPayload


class BlueprintResponse(BaseModel):
    """저장된 blueprint 전체 (blueprint_to_dict 그대로 + 타임스탬프)."""

    model_config = ConfigDict(extra="allow")

    id: str
    name: str
    status: str
    created_at: str
    updated_at: str

    @classmethod
    def from_domain(cls, bp: DocumentBlueprint) -> BlueprintResponse:
        return cls.model_validate(blueprint_to_dict(bp))


class BlueprintSummary(BaseModel):
    id: str
    name: str
    source_kind: str
    page_count: int
    pattern_count: int
    status: str
    updated_at: str

    @classmethod
    def from_domain(cls, bp: DocumentBlueprint) -> BlueprintSummary:
        return cls(
            id=bp.id,
            name=bp.name,
            source_kind=bp.source_kind,
            page_count=bp.page_count,
            pattern_count=len(bp.patterns),
            status=bp.status,
            updated_at=bp.updated_at.isoformat(),
        )


class AssetPreview(BaseModel):
    id: str
    kind: str
    mime: str
    width: int
    height: int
    box: RelBoxSchema
    adopted: bool
    thumbnail_b64: str | None
    # 저장(POST) 시 클라이언트가 되돌려 보내는 원본 바이트 (추출 응답에만 포함)
    data_b64: str | None = None

    @classmethod
    def from_domain(
        cls, a: BlueprintAsset, thumb: str | None, data: bytes | None = None
    ) -> AssetPreview:
        return cls(
            id=a.id,
            kind=a.kind,
            mime=a.mime,
            width=a.width,
            height=a.height,
            box=RelBoxSchema(x=a.box.x, y=a.box.y, w=a.box.w, h=a.box.h),
            adopted=a.adopted,
            thumbnail_b64=thumb,
            data_b64=base64.b64encode(data).decode("ascii") if data else None,
        )


class ClassificationSummary(BaseModel):
    succeeded: int
    failed: int


class BlueprintExtractResponse(BaseModel):
    draft: BlueprintResponse
    assets: list[AssetPreview]
    page_thumbnails: list[str | None]
    classification: ClassificationSummary
    timings_ms: dict[str, int]

    @classmethod
    def from_outcome(
        cls, out: ExtractionOutcome, thumbnailer: Callable[[bytes], str | None]
    ) -> BlueprintExtractResponse:
        bp = out.blueprint
        return cls(
            draft=BlueprintResponse.from_domain(bp),
            assets=[
                AssetPreview.from_domain(
                    a,
                    thumbnailer(out.assets[a.id]) if a.id in out.assets else None,
                    out.assets.get(a.id),
                )
                for a in bp.assets
            ],
            page_thumbnails=[
                thumbnailer(t) if t else None for t in out.page_thumbnails
            ],
            classification=ClassificationSummary(
                succeeded=out.classification[0], failed=out.classification[1]
            ),
            timings_ms=dict(out.timings_ms),
        )


class FontsResponse(BaseModel):
    installed: list[str]
    default: str


class BlueprintOption(BaseModel):
    id: str
    name: str


class BlueprintOptionsResponse(BaseModel):
    items: list[BlueprintOption]
